File: Solution/860.柠檬水找零/860.柠檬水找零.py
'''
Descroption: LeetCode 860. 柠檬水找零
Author: EmoryHuang
Date: 2021-06-15 16:34:00
解题思路:
贪心思想，记录当前5美元和10美元的个数
然后按顺序进行分类讨论
'''


class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        if bills[0] != 5:
            return False
        five, ten = 0, 0
        for bill in bills:
            if bill == 5:
                five += 1
            elif bill == 10:
                if five == 0:
                    return False
                ten += 1
                five -= 1
            else:
                if five > 0 and ten > 0:
                    five -= 1
                    ten -= 1
                elif five >= 3:
                    five -= 3
                else:
                    return False
        return True

File: Solution/860.柠檬水找零/860.柠檬水找零.java
/*
 * @Descroption: LeetCode 860. 柠檬水找零
 * @Author: EmoryHuang
 * @Date: 2023-07-22 09:23:40
 * @Method:
 * 位运算
 */

import java.util.*;

class Solution {
    public boolean lemonadeChange(int[] bills) {
        int cnt5 = 0, cnt10 = 0, cnt20 = 0;
        for (int num : bills) {
            if (num == 5) {
                cnt5++;
            } else if (num == 10) {
                if (cnt5 == 0) {
                    return false;
                }
                cnt10++;
                cnt5--;
            } else if (num == 20) {
                if (cnt5 > 0 && cnt10 > 0) {
                    cnt5--;
                    cnt10--;
                } else if (cnt5 >= 3) {
                    cnt5 -= 3;
                } else {
                    return false;
                }
            }
        }
        return true;
    }
}